### PR TITLE
Restructure parsing flow in the `LightningCLI`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Splitting of some of the changes from #7508

- Breaking changes listed below
- Make `seed_everything` optional
- Typing fixes
- Docs fixes
- Improve config handling

### Does your PR introduce any breaking changes? If yes, please list them.

`LightningCLI` is beta, but these are the non-BC changes

- `init_parser` now returns the parser instance.
- `add_core_arguments_to_parser`, `parse_arguments` now take a `parser` argument
- `instantiate_trainer` now takes a config and a list of callbacks.
- Split `add_core_arguments_to_parser` into `add_default_arguments_to_parser` + `add_core_arguments_to_parser`

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @mauvilsa

### CODE REVIEW

File: pytorch_lightning/utilities/cli.py [hunk 1]
```diff
@@ -325,41 +341,40 @@ def link_optimizers_and_lr_schedulers(self) -> None:
...
-
-    def instantiate_trainer(self) -> None:
-        """Instantiates the trainer using self.config_init['trainer']"""
-        if self.config_init["trainer"].get("callbacks") is None:
-            self.config_init["trainer"]["callbacks"] = []
         callbacks = [self.config_init[c] for c in self.parser.callback_keys]
-        self.config_init["trainer"]["callbacks"].extend(callbacks)
+        self.trainer = self.instantiate_trainer(self.config_init["trainer"], callbacks)
+
+    def instantiate_trainer(self, config: Dict[str, Any], callbacks: List[Callback]) -> Trainer:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/cli.py:359]: callbacks appearing here as an argument seems a bit random. what will happen in the future, will loggers etc. be added here?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/cli.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/utilities/cli.py:359]: Depends.

If you look at the code, you'll see that this list of callbacks is added to the list of callbacks in the trainer config itself.

```python
config["callbacks"] = config["callbacks"] or []
config["callbacks"].extend(callbacks)
```
This is because support for default callbacks defined outside of the trainer config was added in #7964.

With these changes, I'm trying to avoid passing the entire config dictionaries to the `instantiate_trainer` method, but instead just pass the trainer config.

Since these "default" callbacks are not inside the trainer config, they also need to be passed as an input to the method.

So, if in the future we implement the same for loggers, yes we would need to pass them too.

**Review by mauvilsa — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
